src/ORM/Association.php

```diff
@@ -995,11 +995,11 @@ function (CollectionInterface $results, SelectQuery $query) use ($formatters, $p
                     $extracted[] = $result;
                 }
                 $extracted = $query->resultSetFactory()->createResultSet($extracted);
-                $decoratorClass = $query->resultSetFactory()->decoratorClass();
+                $resultSetClass = $query->resultSetFactory()->getResultSetClass();
                 foreach ($formatters as $callable) {
                     $extracted = $callable($extracted, $query);
                     if (!$extracted instanceof ResultSetInterface) {
-                        $extracted = new $decoratorClass($extracted);
+                        $extracted = new $resultSetClass($extracted);
```

Comment:
If a result formatter returns a non-array value, wouldn't using `ResultSet` cause an error where as `ResultSetDecorator` would work?

I think skipping the decorator makes sense if the value is an array. For other iterable types though we should continue using the decorator or we could break userland code.

Reply:
On the `5.next` branch `ResultSet` and `ResultSetDecorator` are almost identical, both simply extend `Collection` (they only have slightly different implementation of `__debugInfo()`). 

So swapping one for another makes no practical difference. I can't foresee a case where dropping use of `ResultSetDecorator` could break userland code.

Reply:
:+1: I was looking at constructors on the 5.x branch 🤦